### PR TITLE
Blog: why histograms?

### PR DESCRIPTION
**Pull request by dyladan**
Guest post from https://dyladan.me/histograms/2023/05/02/why-histograms/

This is the first in a series of 3 posts. The second explains the differences and tradeoffs between histograms and summaries (https://dyladan.me/histograms/2023/05/03/histograms-vs-summaries/). The third explains exponential histograms in detail (https://dyladan.me/histograms/2023/05/04/exponential-histograms/).

Please feel free to share any editing or copy suggestions.

---

**Preview**: https://deploy-preview-2658--opentelemetry.netlify.app/blog/2023/why-histograms/

### CODE REVIEW

**Comment by cartermp** [comment]
FYI - formatting check failed, but since we're probably going to do copy-edits, let's minimize your pain and only have you reformat stuff until after we're satisfied with the content 🙂

**Comment by dyladan** [comment]
> FYI - formatting check failed, but since we're probably going to do copy-edits, let's minimize your pain and only have you reformat stuff until after we're satisfied with the content 🙂

It's just as easy to do it now. Reviews may be easier if things are in the expected format.

File: content/en/blog/2023/why-histograms/index.md
Comment (cartermp) [inline content/en/blog/2023/why-histograms/index.md:15]:
```suggestion
A histogram is a multi-value counter that summarizes the distribution of data points.
For example, a histogram may have 3 counters which count the
occurrences of negative, positive, and zero values respectively. Given a series
of numbers, `3`, `-9`, `7`, `6`, `0`, and `-1`, the histogram would count `2`
negative, `1` zero, and `3` positive values. A single histogram data point is
most commonly represented as a bar chart.
```
slight grammar change and changed "one or more" since people don't use histograms for 1 datapoint, even if it supports it

File: content/en/blog/2023/why-histograms/index.md
Comment (cartermp) [inline content/en/blog/2023/why-histograms/index.md:24]:
```suggestion
The above example has only 3 possible output values, but it is common to
have many more in a single histogram. A real-world application typically exports
a histogram every minute that summarizes a metric for the previous minute.
By using histograms this way, you can study how the distribution of your data changes over time.
```

File: content/en/blog/2023/why-histograms/index.md
Comment (cartermp) [inline content/en/blog/2023/why-histograms/index.md:41]:
```suggestion
A common use-case for histograms in observability is defining
service level objectives (SLOs). One example of such an SLO might be ">=99% of
all queries should respond in less than 30ms," or "90% of all page loads should
become interactive within 100ms of first paint."
```

Comment (cartermp) [reply content/en/blog/2023/why-histograms/index.md:41]: I don't know if this should be caveated here or not (we'd want to do that in docs though). This point is true when you're working on a metrics data store, but it won't necessarily be true for other tools that let you establish an SLO based on raw data from traces or logs. I'll leave that up to you if you wanna add it or not; I don't think it's urgent to caveat.

File: content/en/blog/2023/why-histograms/index.md
Comment (cartermp) [inline content/en/blog/2023/why-histograms/index.md:36]:
```suggestion
## Why are Histograms useful?
```

File: content/en/blog/2023/why-histograms/index.md
Comment (cartermp) [inline content/en/blog/2023/why-histograms/index.md:26]:
```suggestion
## What are histograms for?
```

File: content/en/blog/2023/why-histograms/index.md
Comment (cartermp) [inline content/en/blog/2023/why-histograms/index.md:53]:
```suggestion
## Other metric types
```

File: content/en/blog/2023/why-histograms/index.md
Comment (cartermp) [inline content/en/blog/2023/why-histograms/index.md:77]:
```suggestion
## Other data sources and metric types
```

File: content/en/blog/2023/why-histograms/index.md
Comment (cartermp) [inline content/en/blog/2023/why-histograms/index.md:66]:
```suggestion
What if you're already defining SLOs based on other metrics?
You may have considered defining the SLOs to be based on gauges or counters.
This approach can work, but it requires defining your SLOs before understand your
data distribution and requires non-trivial implementation at collection time. It
is also inflexible; if you decide to change your SLO from 90% of requests to 99%
of requests, you have to make and release code changes, then wait for the old
data to age out and the new metric to collect enough data to make useful
queries. Because histograms model data as a distribution from start to finish,
they enable you to simply change your queries and get answers on the data you've
already collected. Particularly with exponential histograms, arbitrary
distribution queries can be made with very low relative error rates and minimal
resource consumption on both the client and the analysis backend.
```

File: content/en/blog/2023/why-histograms/index.md
Comment (cartermp) [inline content/en/blog/2023/why-histograms/index.md:75]:
```suggestion
The inflexibility of not using histograms for SLOs also impacts your ability to gauge impact
when your SLO is violated. For example, imagine you are collecting a gauge that
calculates the `p99` of some metric and you define an SLO based on it. When your
SLO is violated and an alert is triggered, how do you know it is really only
affecting 1% of queries, 10%, or 50%? A histogram allows you to answer that
question by querying the percentiles you're interested in. You could technically
collect additional gauges for each percentile, but that's just an ad-hoc reimplementation
of histograms anyways, so you're better off using histograms.
```

File: content/en/blog/2023/why-histograms/index.md
Comment (cartermp) [inline content/en/blog/2023/why-histograms/index.md:86]:
```suggestion
You may ask, "why would you report a separate metric rather than calculating these
metrics from your existing log and trace data? While it is true that for _some_
use cases, like response times, this may be possible, it is not necessarily possible for
_all_ use cases. Even when percentiles can be calculated from existing data,
this typically requires querying a massive amount of data to answer a very
simple question. Some observability backends can support this, but others cannot.
Further, if you are sampling your tracing and logs data, you miss crucial information
required to asses whether or not you are meeting your SLOs. Again, this concern is
also backend- and sampling-specific. Depending on your sampling strategy, it may
not be possible to have representative data to accurately calculate an SLO. Further,
if your backend doesn't support re-weighting counts on sampled data, it can
similarly lead to inaccurate SLOs. Using histograms is a way to avoid these subtle
problems if they apply to you.
```

**Review by cartermp — COMMENTED** [review]
Looks great overall, thank you! Mostly some small copy-edits, but I modified the last paragraph a a bit to caveat about backends and sampling strategies. Since accurate calculations of SLOs is a super deep topic we probably shouldn't elaborate too much, but I do think we should call out that the problems you can run into won't be the same from backend to backend.

File: content/en/blog/2023/why-histograms/index.md
Comment (dyladan) [reply content/en/blog/2023/why-histograms/index.md:26]: Every heading is H2? Is that a style thing for the Otel blog in general?

File: content/en/blog/2023/why-histograms/index.md
Comment (dyladan) [reply content/en/blog/2023/why-histograms/index.md:39]: >This point is true when you're working on a metrics data store, but it won't necessarily be true for other tools that let you establish an SLO based on raw data from traces or logs.

Did you mean to put this on a different line? I'm not sure what claim here might be untrue. It is a common use case for histograms. In any case I like the specific copy edit suggestion you made. "The most" is a bit strong

File: content/en/blog/2023/why-histograms/index.md
Comment (dyladan) [reply content/en/blog/2023/why-histograms/index.md:84]: What about this:

>You may ask, "why would I report a separate metric rather than calculating it
from my existing log and trace data?" While it is true that for _some_ use
cases, like response times, this may be possible, it is not necessarily possible
for _all_ use cases. Even when quantiles can be calculated from existing data,
you may run into other problems. You need to be sure your observability backend
is able to query and analyze a large amount of existing data on-line or index
and analyze it at ingestion time. If you are sampling your logs and traces or
employing a data retention policy that ages data out, you need to be sure those
things are not affecting derived metrics, or that they are properly re-weighted,
or you risk not being able to accurately asses your SLOs. Depending on your
sampling strategy, it may not even be possible. Using histograms is a way to
avoid these subtle problems if they apply to you.

Comment (dyladan) [reply content/en/blog/2023/why-histograms/index.md:84]: I went out of my way when writing the whole series to try not to mention anything backend specific so I'd like to focus the wording on the underlying principles instead.

Comment (cartermp) [reply content/en/blog/2023/why-histograms/index.md:84]: I love that @dyladan, thanks! Let's go with that and then the post LGTM.

File: content/en/blog/2023/why-histograms/index.md
Comment (chalin) [inline content/en/blog/2023/why-histograms/index.md:8]:
```suggestion
canonical_url: https://dyladan.me/histograms/2023/05/02/why-histograms/
---

```

Comment (chalin) [reply content/en/blog/2023/why-histograms/index.md:8]: To be consistent with our other posts, please add the following at the end of the page:

```markdown
_A version of this article was [originally posted][] to the author's blog._

[originally posted]: {{% param canonical_url %}}
```

**Review by chalin — COMMENTED** [review]
See inline comments, thanks.

File: content/en/blog/2023/why-histograms/index.md
Comment (chalin) [inline content/en/blog/2023/why-histograms/index.md:49]:
```suggestion
![p99, p90, and p50 plotted as lines](hist-lines.png "p99, p90, and p50 plotted as a line chart with title 'response times.' Time is on the x-axis and response times in milliseconds on the y-axis. p99 response times are around 80 milliseconds. p90 response times are between 60 and 80 milliseconds. p50 response times are between 20 and 30 milliseconds.")
```

File: content/en/blog/2023/why-histograms/index.md
Comment (chalin) [inline content/en/blog/2023/why-histograms/index.md:4]:
```suggestion
author: '[Daniel Dyla](https://github.com/dyladan)'
spelling: cSpell:ignore Dyla quantile quantiles reimplementation
```

**Comment by dyladan** [comment]
updated the date to monday may 8th to reflect when this is likely to be merged

**Comment by dyladan** [comment]
If a maintainer merges this before that obviously they can change the date. I have enabled "allow edits from maintainers" on the PR

**Comment by dyladan** [comment]
@cartermp @chalin I rewrote the section "Other metric types" because on re-reading I noticed that the wording was technically incorrect. The last paragraph previously stated that collecting each percentile as a gauge is an ad-hoc implementation of histograms, when this is actually much closer to a summary metric type. I also added a caveat that summaries cannot be aggregated across hosts.

**Review by cartermp — APPROVED** [review]
Thanks! Great content. Unless there's no other objections, I'll merge this on Monday and it'll be live.

**Review by chalin — APPROVED** [review]
👍

**Pull request merged**
